code/ship/ship.cpp

```diff
@@ -11902,7 +11902,12 @@ void change_ship_type(int n, int ship_type, int by_sexp)
 
 	// zookeeper - If we're switching in the loadout screen, make sure we retain initial velocity set in FRED
 	if (!(Game_mode & GM_IN_MISSION) && !(Fred_running)) {
-		Objects[sp->objnum].phys_info.speed = (float) p_objp->initial_velocity * sip->max_speed / 100.0f;
+		if (p_objp != nullptr){
```

Comment:
The nullptr check should be part of the enclosing if().  If there is no parse object, there is no need to change the velocity at all, and it should not be set to 0.  And if the velocity isn't changed, the rest of the if() block is redundant.

Reply:
It's conceivable that your assumption could be invalidated.  Ship lab code could call this in the future, or some other block of code.  So it's better for this to be accessible for odd cases where we wouldn't have a parse object.

Reply:
The initial velocity is a property of the parse object.  That's not an assumption, that's a fact of the code.  If there is no parse object, it is meaningless to talk about the initial velocity.  Therefore, if there is no parse object, all of the initial velocity-related code can be skipped.

Suppose a ship is created via ship-create.  In that situation it has no parse object.  So if change-ship-class is called on it, the speed should not be set to 0; the speed should be left alone.